File: services/api/api/routers/crm.py
"""CRM endpoints: pipeline stages, leads, and activity log."""
from __future__ import annotations

from typing import Annotated, Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel

from api.auth import get_merchant_id
from api.db import postgres
from api.limiter import limiter

router = APIRouter(prefix="/api/v1/crm")

VALID_ACTIVITY_TYPES = {"note", "email", "call", "meeting", "stage_change"}


# ── Schemas ───────────────────────────────────────────────────────────────────

class StageCreate(BaseModel):
    name: str
    color: str = "#6B7280"


class StageUpdate(BaseModel):
    name: Optional[str] = None
    color: Optional[str] = None
    position: Optional[int] = None


class LeadCreate(BaseModel):
    name: str
    email: Optional[str] = None
    company: Optional[str] = None
    phone: Optional[str] = None
    stage_id: Optional[str] = None
    plan_interest: Optional[str] = None
    mrr_estimate_paise: int = 0
    source: Optional[str] = None
    owner: Optional[str] = None
    notes: Optional[str] = None


class LeadUpdate(BaseModel):
    name: Optional[str] = None
    email: Optional[str] = None
    company: Optional[str] = None
    phone: Optional[str] = None
    stage_id: Optional[str] = None
    plan_interest: Optional[str] = None
    mrr_estimate_paise: Optional[int] = None
    source: Optional[str] = None
    owner: Optional[str] = None
    notes: Optional[str] = None


class ActivityCreate(BaseModel):
    type: str = "note"
    body: str


class TaskCreate(BaseModel):
    lead_id: Optional[str] = None
    title: str
    description: Optional[str] = None
    assignee: Optional[str] = None
    due_date: Optional[str] = None


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    assignee: Optional[str] = None
    due_date: Optional[str] = None
    status: Optional[str] = None


class SequenceCreate(BaseModel):
    name: str


class StepCreate(BaseModel):
    step_num: int
    delay_days: int = 0
    subject: str
    body: str


class EnrollRequest(BaseModel):
    lead_id: str


# ── Stages ────────────────────────────────────────────────────────────────────

@router.get("/stages")
@limiter.limit("60/minute")
def list_stages(
    request: Request,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return {"stages": postgres.list_pipeline_stages(merchant_id)}


@router.post("/stages", status_code=201)
@limiter.limit("30/minute")
def create_stage(
    request: Request,
    body: StageCreate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return postgres.create_pipeline_stage(merchant_id, body.name, body.color)


@router.put("/stages/{stage_id}")
@limiter.limit("60/minute")
def update_stage(
    request: Request,
    stage_id: str,
    body: StageUpdate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    updated = postgres.update_pipeline_stage(
        merchant_id, stage_id, body.model_dump(exclude_none=True)
    )
    if not updated:
        raise HTTPException(status_code=404, detail="stage not found")
    return updated


@router.delete("/stages/{stage_id}", status_code=204)
@limiter.limit("30/minute")
def delete_stage(
    request: Request,
    stage_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    ok = postgres.delete_pipeline_stage(merchant_id, stage_id)
    if not ok:
        raise HTTPException(status_code=409, detail="stage has leads or was not found")


# ── Leads ─────────────────────────────────────────────────────────────────────

@router.get("/leads")
@limiter.limit("60/minute")
def list_leads(
    request: Request,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
    stage_id: Optional[str] = None,
):
    return {"leads": postgres.list_crm_leads(merchant_id, stage_id)}


@router.post("/leads", status_code=201)
@limiter.limit("30/minute")
def create_lead(
    request: Request,
    body: LeadCreate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return postgres.create_crm_lead(merchant_id, body.model_dump())


@router.get("/leads/{lead_id}")
@limiter.limit("60/minute")
def get_lead(
    request: Request,
    lead_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    lead = postgres.get_crm_lead(merchant_id, lead_id)
    if not lead:
        raise HTTPException(status_code=404, detail="lead not found")
    activities = postgres.list_lead_activities(merchant_id, lead_id)
    return {**lead, "activities": activities}


@router.put("/leads/{lead_id}")
@limiter.limit("60/minute")
def update_lead(
    request: Request,
    lead_id: str,
    body: LeadUpdate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    updated = postgres.update_crm_lead(
        merchant_id, lead_id, body.model_dump(exclude_none=True)
    )
    if not updated:
        raise HTTPException(status_code=404, detail="lead not found")
    return updated


@router.delete("/leads/{lead_id}", status_code=204)
@limiter.limit("30/minute")
def delete_lead(
    request: Request,
    lead_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    ok = postgres.delete_crm_lead(merchant_id, lead_id)
    if not ok:
        raise HTTPException(status_code=404, detail="lead not found")


@router.post("/leads/{lead_id}/activities", status_code=201)
@limiter.limit("60/minute")
def add_activity(
    request: Request,
    lead_id: str,
    body: ActivityCreate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    if body.type not in VALID_ACTIVITY_TYPES:
        raise HTTPException(status_code=422, detail=f"type must be one of {sorted(VALID_ACTIVITY_TYPES)}")
    lead = postgres.get_crm_lead(merchant_id, lead_id)
    if not lead:
        raise HTTPException(status_code=404, detail="lead not found")
    return postgres.add_lead_activity(merchant_id, lead_id, body.type, body.body)


# ── Pipeline (full board snapshot) ───────────────────────────────────────────

@router.get("/pipeline")
@limiter.limit("60/minute")
def get_pipeline(
    request: Request,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    stages = postgres.list_pipeline_stages(merchant_id)
    leads = postgres.list_crm_leads(merchant_id)

    leads_by_stage: dict[str, list] = {}
    for lead in leads:
        key = lead.get("stage_id") or "__none__"
        leads_by_stage.setdefault(key, []).append(lead)

    pipeline = [{**s, "leads": leads_by_stage.get(s["id"], [])} for s in stages]
    return {"pipeline": pipeline, "unassigned": leads_by_stage.get("__none__", [])}


# ── Tasks ─────────────────────────────────────────────────────────────────────

@router.get("/tasks")
@limiter.limit("60/minute")
def list_tasks(
    request: Request,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
    lead_id: Optional[str] = None,
    status: Optional[str] = None,
):
    return {"tasks": postgres.list_tasks(merchant_id, lead_id, status)}


@router.post("/tasks", status_code=201)
@limiter.limit("30/minute")
def create_task(
    request: Request,
    body: TaskCreate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return postgres.create_task(merchant_id, body.model_dump())


@router.put("/tasks/{task_id}")
@limiter.limit("60/minute")
def update_task(
    request: Request,
    task_id: str,
    body: TaskUpdate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    updated = postgres.update_task(merchant_id, task_id, body.model_dump(exclude_none=True))
    if not updated:
        raise HTTPException(status_code=404, detail="task not found")
    return updated


@router.delete("/tasks/{task_id}", status_code=204)
@limiter.limit("30/minute")
def delete_task(
    request: Request,
    task_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    ok = postgres.delete_task(merchant_id, task_id)
    if not ok:
        raise HTTPException(status_code=404, detail="task not found")


# ── Sequences ─────────────────────────────────────────────────────────────────

@router.get("/sequences")
@limiter.limit("60/minute")
def list_sequences(
    request: Request,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return {"sequences": postgres.list_sequences(merchant_id)}


@router.post("/sequences", status_code=201)
@limiter.limit("20/minute")
def create_sequence(
    request: Request,
    body: SequenceCreate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return postgres.create_sequence(merchant_id, body.name)


@router.get("/sequences/{seq_id}")
@limiter.limit("60/minute")
def get_sequence(
    request: Request,
    seq_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    seq = postgres.get_sequence(merchant_id, seq_id)
    if not seq:
        raise HTTPException(status_code=404, detail="sequence not found")
    return seq


@router.delete("/sequences/{seq_id}", status_code=204)
@limiter.limit("20/minute")
def delete_sequence(
    request: Request,
    seq_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    ok = postgres.delete_sequence(merchant_id, seq_id)
    if not ok:
        raise HTTPException(status_code=404, detail="sequence not found")


@router.post("/sequences/{seq_id}/steps", status_code=201)
@limiter.limit("30/minute")
def add_step(
    request: Request,
    seq_id: str,
    body: StepCreate,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return postgres.add_sequence_step(merchant_id, seq_id, body.step_num, body.delay_days, body.subject, body.body)


@router.delete("/sequences/{seq_id}/steps/{step_id}", status_code=204)
@limiter.limit("20/minute")
def delete_step(
    request: Request,
    seq_id: str,
    step_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    ok = postgres.delete_sequence_step(merchant_id, seq_id, step_id)
    if not ok:
        raise HTTPException(status_code=404, detail="step not found")


@router.post("/sequences/{seq_id}/enroll", status_code=201)
@limiter.limit("30/minute")
def enroll(
    request: Request,
    seq_id: str,
    body: EnrollRequest,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    lead = postgres.get_crm_lead(merchant_id, body.lead_id)
    if not lead:
        raise HTTPException(status_code=404, detail="lead not found")
    return postgres.enroll_lead(merchant_id, seq_id, body.lead_id)


@router.delete("/sequences/{seq_id}/enroll/{lead_id}", status_code=204)
@limiter.limit("20/minute")
def unenroll(
    request: Request,
    seq_id: str,
    lead_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    postgres.unenroll_lead(merchant_id, seq_id, lead_id)


@router.get("/leads/{lead_id}/enrollments")
@limiter.limit("60/minute")
def lead_enrollments(
    request: Request,
    lead_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return {"enrollments": postgres.list_lead_enrollments(merchant_id, lead_id)}


# ── Rep Performance ───────────────────────────────────────────────────────────

@router.get("/reps")
@limiter.limit("30/minute")
def rep_stats(
    request: Request,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    return {"reps": postgres.get_rep_stats(merchant_id)}


# ── Lead Enrichment ───────────────────────────────────────────────────────────

@router.post("/leads/{lead_id}/enrich")
@limiter.limit("20/minute")
def enrich_lead(
    request: Request,
    lead_id: str,
    merchant_id: Annotated[str, Depends(get_merchant_id)],
):
    enriched = postgres.enrich_lead(merchant_id, lead_id)
    if not enriched:
        raise HTTPException(status_code=404, detail="lead not found")
    return enriched

File: services/frontend/src/api/crm.ts
import { createBearerClient } from "./client";

export interface Stage {
  id: string;
  name: string;
  position: number;
  color: string;
}

export interface Lead {
  id: string;
  stage_id: string | null;
  customer_id: string | null;
  name: string;
  email: string | null;
  company: string | null;
  phone: string | null;
  plan_interest: string | null;
  mrr_estimate_paise: number;
  source: string | null;
  owner: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface Activity {
  id: string;
  type: string;
  body: string;
  created_at: string;
}

export interface LeadWithActivities extends Lead {
  activities: Activity[];
}

export interface StageWithLeads extends Stage {
  leads: Lead[];
}

export interface PipelineResponse {
  pipeline: StageWithLeads[];
  unassigned: Lead[];
}

// ── Stages ────────────────────────────────────────────────────────────────────

export async function fetchStages(token: string): Promise<Stage[]> {
  const { data } = await createBearerClient(token).get("/api/v1/crm/stages");
  return data.stages;
}

export async function createStage(token: string, name: string, color?: string): Promise<Stage> {
  const { data } = await createBearerClient(token).post("/api/v1/crm/stages", { name, color });
  return data;
}

export async function updateStage(
  token: string,
  stageId: string,
  patch: Partial<Pick<Stage, "name" | "color" | "position">>,
): Promise<Stage> {
  const { data } = await createBearerClient(token).put(`/api/v1/crm/stages/${stageId}`, patch);
  return data;
}

export async function deleteStage(token: string, stageId: string): Promise<void> {
  await createBearerClient(token).delete(`/api/v1/crm/stages/${stageId}`);
}

// ── Leads ─────────────────────────────────────────────────────────────────────

export async function fetchLeads(token: string, stageId?: string): Promise<Lead[]> {
  const params = stageId ? { stage_id: stageId } : {};
  const { data } = await createBearerClient(token).get("/api/v1/crm/leads", { params });
  return data.leads;
}

export async function fetchPipeline(token: string): Promise<PipelineResponse> {
  const { data } = await createBearerClient(token).get("/api/v1/crm/pipeline");
  return data;
}

export async function fetchLead(token: string, leadId: string): Promise<LeadWithActivities> {
  const { data } = await createBearerClient(token).get(`/api/v1/crm/leads/${leadId}`);
  return data;
}

export async function createLead(
  token: string,
  lead: Omit<Lead, "id" | "customer_id" | "created_at" | "updated_at">,
): Promise<Lead> {
  const { data } = await createBearerClient(token).post("/api/v1/crm/leads", lead);
  return data;
}

export async function updateLead(
  token: string,
  leadId: string,
  patch: Partial<Omit<Lead, "id" | "customer_id" | "created_at" | "updated_at">>,
): Promise<Lead> {
  const { data } = await createBearerClient(token).put(`/api/v1/crm/leads/${leadId}`, patch);
  return data;
}

export async function deleteLead(token: string, leadId: string): Promise<void> {
  await createBearerClient(token).delete(`/api/v1/crm/leads/${leadId}`);
}

export async function addActivity(
  token: string,
  leadId: string,
  type: string,
  body: string,
): Promise<Activity> {
  const { data } = await createBearerClient(token).post(`/api/v1/crm/leads/${leadId}/activities`, { type, body });
  return data;
}

export async function enrichLead(token: string, leadId: string): Promise<Lead> {
  const { data } = await createBearerClient(token).post(`/api/v1/crm/leads/${leadId}/enrich`);
  return data;
}

// ── Tasks ─────────────────────────────────────────────────────────────────────

export interface Task {
  id: string;
  lead_id: string | null;
  title: string;
  description: string | null;
  assignee: string | null;
  due_date: string | null;
  status: "open" | "done";
  created_at: string;
  updated_at: string;
}

export async function fetchTasks(token: string, leadId?: string, status?: string): Promise<Task[]> {
  const params: Record<string, string> = {};
  if (leadId) params.lead_id = leadId;
  if (status) params.status = status;
  const { data } = await createBearerClient(token).get("/api/v1/crm/tasks", { params });
  return data.tasks;
}

export async function createTask(
  token: string,
  task: { title: string; lead_id?: string; description?: string; assignee?: string; due_date?: string },
): Promise<Task> {
  const { data } = await createBearerClient(token).post("/api/v1/crm/tasks", task);
  return data;
}

export async function updateTask(
  token: string,
  taskId: string,
  patch: Partial<Pick<Task, "title" | "description" | "assignee" | "due_date" | "status">>,
): Promise<Task> {
  const { data } = await createBearerClient(token).put(`/api/v1/crm/tasks/${taskId}`, patch);
  return data;
}

export async function deleteTask(token: string, taskId: string): Promise<void> {
  await createBearerClient(token).delete(`/api/v1/crm/tasks/${taskId}`);
}

// ── Sequences ─────────────────────────────────────────────────────────────────

export interface SequenceStep {
  id: string;
  step_num: number;
  delay_days: number;
  subject: string;
  body: string;
}

export interface Sequence {
  id: string;
  name: string;
  created_at: string;
  step_count: number;
  active_enrollments: number;
}

export interface SequenceDetail extends Sequence {
  steps: SequenceStep[];
}

export interface Enrollment {
  id: string;
  sequence_id: string;
  sequence_name: string;
  current_step: number;
  status: "active" | "completed" | "stopped";
  enrolled_at: string;
}

export interface RepStat {
  rep: string;
  total_leads: number;
  won_leads: number;
  lost_leads: number;
  pipeline_mrr_paise: number;
  new_leads_30d: number;
}

export async function fetchSequences(token: string): Promise<Sequence[]> {
  const { data } = await createBearerClient(token).get("/api/v1/crm/sequences");
  return data.sequences;
}

export async function createSequence(token: string, name: string): Promise<Sequence> {
  const { data } = await createBearerClient(token).post("/api/v1/crm/sequences", { name });
  return data;
}

export async function getSequence(token: string, seqId: string): Promise<SequenceDetail> {
  const { data } = await createBearerClient(token).get(`/api/v1/crm/sequences/${seqId}`);
  return data;
}

export async function deleteSequence(token: string, seqId: string): Promise<void> {
  await createBearerClient(token).delete(`/api/v1/crm/sequences/${seqId}`);
}

export async function addSequenceStep(
  token: string,
  seqId: string,
  step: { step_num: number; delay_days: number; subject: string; body: string },
): Promise<SequenceStep> {
  const { data } = await createBearerClient(token).post(`/api/v1/crm/sequences/${seqId}/steps`, step);
  return data;
}

export async function deleteSequenceStep(token: string, seqId: string, stepId: string): Promise<void> {
  await createBearerClient(token).delete(`/api/v1/crm/sequences/${seqId}/steps/${stepId}`);
}

export async function enrollLead(token: string, seqId: string, leadId: string): Promise<Enrollment> {
  const { data } = await createBearerClient(token).post(`/api/v1/crm/sequences/${seqId}/enroll`, { lead_id: leadId });
  return data;
}

export async function unenrollLead(token: string, seqId: string, leadId: string): Promise<void> {
  await createBearerClient(token).delete(`/api/v1/crm/sequences/${seqId}/enroll/${leadId}`);
}

export async function fetchLeadEnrollments(token: string, leadId: string): Promise<Enrollment[]> {
  const { data } = await createBearerClient(token).get(`/api/v1/crm/leads/${leadId}/enrollments`);
  return data.enrollments;
}

export async function fetchRepStats(token: string): Promise<RepStat[]> {
  const { data } = await createBearerClient(token).get("/api/v1/crm/reps");
  return data.reps;
}
